File: 数据结构 以及部分基础模板算法/并查集/200. 岛屿数量.py

class BQSet():
    def __init__(self):
        self.f = {}  # f[i]代表i的父节点
        # #init
        # for i in range(10):
        #     f[i] = i

    def getFather(self,origin):
        a = origin
        while self.f[a] != a:
            a = self.f[a]
        self.f[origin] = a  # 优化一下
        return a

    # 只需要看a， b 是否 有共同父节点
    def judge(self,a, b):
        a  = self.getFather(a)
        b = self.getFather(b)

        if self.f[a] == self.f[b]:
            return True
        else:
            return False

    def Union(self,source, a):
        a = self.getFather(a)
        sF = self.getFather(source)
        self.f[a] = sF




class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        dx = [1, -1, 0, 0]
        dy = [0, 0, 1, -1]
        l1 = len(grid)
        l2 = len(grid[0])
        bq = BQSet()
        for i in range(l1):
            for j in range(l2):
                now = str(i) + '&' + str(j)
                bq.f[now] = now


        for i in range(l1):
            for j in range(l2):
                now = str(i) + '&' + str(j)
                if grid[i][j] == '1':
                    #分别向四个方向
                    for index in range(4):
                        x = i + dx[index]
                        y = j + dy[index]
                        new  = str(x) + '&' + str(y)
                        try:
                            # 同样就合并
                            if grid[x][y] == '1':
                                bq.Union(now, new)
                        except:
                            pass

        f = {}
        for i in bq.f:
            father = bq.getFather(i)
            x, y = i.split('&')
            x= int(x)
            y = int(y)
            # print(i,father)
            if father not  in f and grid[x][y] == '1':
                f[father] = True

        # print(len(f))
        return len(f)
s  = Solution()
s.numIslands([["1","1","1","1","0"],["1","1","0","1","0"],["1","1","0","0","0"],["0","0","0","0","0"]])

File: 图/other/200. 岛屿数量.java
class Solution {
    public int numIslands(char[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        int ans =0;
        for(int i = 0; i < m ; i++){
            for(int j =0; j < n; j++){
                if (grid[i][j] == '1') {
                    dfs(grid, i, j);
                    ans++;
                    System.out.println("ans: " + ans);
                    System.out.println("i: " + i + " j: " + j);
                }
            }
        }

        return ans;
        
    }

    public void dfs(char[][] grid, int x, int y){
        int m = grid.length;
        int n = grid[0].length;

        
        if (x >= 0 && x < m && y >= 0 && y < n && grid[x][y] == '1') {
            // if(grid[x][y] == '0'){
            //     return;
            // }
            System.out.println("x: " + x + " y: " + y);

            grid[x][y] = '0';
            dfs(grid, x, y-1);
            dfs(grid, x, y+1);
            dfs(grid, x+1, y);
            dfs(grid, x-1, y);
        }
        
    }

    // [["1","1","1","1","0"],["1","1","0","1","0"],["1","1","0","0","0"],["0","0","0","0","0"]]
    public static void main(String[] args) {
        Solution s = new Solution();
        char[][] grid = new char[][]{
            {'1','1','1','1','0'},
            {'1','1','0','1','0'},
            {'1','1','0','0','0'},
            {'0','0','0','0','0'}
        };
        System.out.println(s.numIslands(grid));
    }

}